### ISSUE TITLE
Pipeline configuration in JSON - configrepo extension and new API

### ISSUE THREAD

**Issue by tomzo** [gocd/gocd#1439]
As part of #1133 there is a new extension point `configrepo` that uses JSON to represent pipelines configuration (in plugin-server communication). 
At the same time there is emerging work in #1417 to introduce new JSON server API for configuring pipelines.
- There is a risk that these JSONs will be inconsistent. 
- And there is a risk that someone will implement serialization/deserialization that is already done in (not merged yet) https://github.com/tomzo/gocd/compare/1133-minimal-pr...tomzo:1133-extension . The whole tree of pipeline config is quite big so it would be a waste of efforts.

I think that **from the users perspective it is important to have both APIs as similar as possible**. E.g. both should replace confusing `isLocked` keyword by the same more, sensible alternative.
- We should decide upfront how JSON in new API should look like. We can have discussion in this issue
- Once API is decided or partially implemented the 'schemas' in configrepo should be updated
### Reuse

Some serialization/deserialization code could be reused in new server API once extension point is merged.

The schema in configrepo extension **cannot be entirely identical**:
- I skipped support for parameters and pipeline templates
- new server API is planned without pipeline groups and environments which I implemented as part of #1133

Elements which I think will/should be **identical**:
- tasks
- materials
- jobs
- stages
- environment variables
  Only pipeline config should be different IMO - because of params and templates that I skipped.

Even if we go for HAL+JSON in server API then there are still **elementary** JSON objects in configrepo extension that can be reused. (The ones without references) - some materials, some tasks, approval, environment variable etc.
### Documentation

Another thing is that developer of configrepo plugin does not have any documentation with pipeline config schema at the moment. Documentations would be similar as well.
I guess it would be nice to have configrepo docs in same new-style like http://api.go.cd/current/
### Questions
- How are you planning to solve references like `scmId` in [PluggableSCMMaterialConfig](https://github.com/gocd/gocd/blob/f52a72f9b7c23919470dbdbc47139ef1dc95acf8/config/config-api/src/com/thoughtworks/go/config/materials/PluggableSCMMaterialConfig.java#L45)?
- How to solve migration/ changing pipeline schema? Will old API client still work after server upgrade?
- Is it time to create actual JSON schema of the pipelines (equivalent of part of cruise-config.xsd)? Considering #866
